src/PackageManagement/BuildIntegration/BuildIntegratedRestoreUtility.cs

```diff
@@ -109,5 +141,155 @@ public static IReadOnlyList<PackageIdentity> GetRemovedPackages(LockFile origina
         {
             return GetAddedPackages(updatedLockFile, originalLockFile);
         }
+
+        /// <summary>
+        /// Creates an index of the project unique name to the cache entry. 
+        /// The cache entry contains the project and the closure of project.json files.
+        /// </summary>
+        public static async Task<IReadOnlyDictionary<string, BuildIntegratedProjectCacheEntry>> 
+            CreateBuildIntegratedProjectStateCache(IReadOnlyList<BuildIntegratedNuGetProject> projects)
+        {
+            var cache = new Dictionary<string, BuildIntegratedProjectCacheEntry>();
+
+            // Find all project closures
+            foreach (var project in projects)
+            {
+                // Get all project.json file paths in the closure
+                var closure = await project.GetProjectReferenceClosureAsync();
+                var files = closure.Select(reference => reference.PackageSpecPath).ToList();
+
+                var projectInfo = new BuildIntegratedProjectCacheEntry(project.JsonConfigPath, files);
+
+                var uniqueName = project.GetMetadata<string>(NuGetProjectMetadataKeys.UniqueName);
```

Comment:
I would recommend using NuGetProject.GetUniqueNameOrName method which will not throw if the unique name is not available and use the name